### PR TITLE
Swarm 720 - Update Logging Fraction to support the creation of a default File Logging handler

### PR DESCRIPTION
**Pull request by cmoulliard**
- Update Logging Fraction to support the creation of a default File Logging handler
- Add new default values : DEFAULT_FILE_HANDLER_NAME, DEFAULT_LOGGING_FILE_NAME, DEFAULT_LOGGING_DIR
- Design new test cases to support the DSL
- Include package name to also configure the Logger with the level passed

### CODE REVIEW

File: testsuite/testsuite-logging/src/test/java/org/wildfly/swarm/logging/DefaultFileLoggingArquillianTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Test
+    public void doLogging() throws FileNotFoundException {
+        String message = "testing: " + UUID.randomUUID().toString();
+        Logger logger = Logger.getLogger("org.wildfly.swarm.logging");
+        logger.info(message);
+        assertTrue("File not found: " + logFile, new File(logFile).exists());
+
+        BufferedReader reader = new BufferedReader(new FileReader(logFile));
+        List<String> lines = reader.lines().collect(Collectors.toList());
```
Comment (gastaldi) [inline testsuite/testsuite-logging/src/test/java/org/wildfly/swarm/logging/DefaultFileLoggingArquillianTest.java]: You need to close this reader or use Files.readAllLines(Path) instead

**Comment by gastaldi** [comment]
Add to whitelist

**Review by heiko-braun — APPROVED** [review]

**Review by gastaldi — APPROVED** [review]

**Comment by bobmcwhirter** [comment]
retest this please

**Comment by bobmcwhirter** [comment]
Would you mind squashing and rebasing this before we merge it?

**Review by bobmcwhirter — COMMENTED** [review]
This seems to improve things if you have exactly 1 file and 1 package you want logged to it.

Do you reckon we could do something non-static on LoggingFraction to make setting up N files for N packages easier?  Would be chainable from existing logging-fractions setup however.

e.g.

```
myLoggingFraction.logToFile( "foo.txt", "com.foo" ).logToFile( "bar.txt", "org.bar" )
```

etc

**Comment by cmoulliard** [comment]
> Would you mind squashing and rebasing this before we merge it?

No problem.

**Comment by cmoulliard** [comment]
> do something non-static on LoggingFraction

We can do that for sure but that should maybe done globally as all the factory methods are defined statically (createDefaultLoggingFraction, createErrorLoggingFraction, createDebugLoggingFraction, ...) ?

**Comment by cmoulliard** [comment]
> Do you reckon we could do something non-static on LoggingFraction to make setting up N files for N packages easier?

I will work on that and refactor LoggingFraction to support the chaining

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java [hunk 2]
```diff
@@ -81,7 +87,7 @@ public LoggingFraction applyDefaults(Level level) {
      *
      * @return The fully-configured fraction.
      */
-    public static LoggingFraction createTraceLoggingFraction() {
```
Comment (gastaldi) [inline logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:84]: Removing static here would break backward compatibilty

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java [hunk 3]
```diff
@@ -90,7 +96,7 @@ public static LoggingFraction createTraceLoggingFraction() {
      *
      * @return The fully-configured fraction.
      */
-    public static LoggingFraction createDebugLoggingFraction() {
+    public LoggingFraction createDebugLoggingFraction() {
```
Comment (bobmcwhirter) [inline logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java]: This should remain static.

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java [hunk 4]
```diff
@@ -99,7 +105,7 @@ public static LoggingFraction createDebugLoggingFraction() {
      *
      * @return The fully-configured fraction.
      */
-    public static LoggingFraction createErrorLoggingFraction() {
+    public LoggingFraction createErrorLoggingFraction() {
```
Comment (bobmcwhirter) [inline logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java]: As should this.  I think something is decided askew.

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java (hunk 3 above)
Comment (cmoulliard) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java]: I will switch to use static method & resubmit

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java (hunk 2 above)
Comment (cmoulliard) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:84]: No problem

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java [hunk 5]
```diff
@@ -51,6 +60,9 @@
 
     public static final String COLOR_PATTERN = "COLOR_PATTERN";
 
+    public static LoggingFraction loggingFraction;
```
Comment (gastaldi) [inline logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:63]: Why have a static reference here?

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java [hunk 6]
```diff
@@ -76,6 +88,13 @@ public LoggingFraction applyDefaults(Level level) {
         return this;
     }
 
+    public static LoggingFraction getInstance() {
```
Comment (gastaldi) [inline logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:91]: I don't think this is needed due to the existing static createXX() methods

**Review by gastaldi — CHANGES_REQUESTED** [review]

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java (hunk 5 above)
Comment (cmoulliard) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:63]: To have one singleton / class of the LoggingFraction. Otherwise, we can't chain the methods.

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java (hunk 6 above)
Comment (cmoulliard) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:91]: To have one singleton / class of the LoggingFraction. Otherwise, we can't chain the methods.

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java (hunk 5 above)
Comment (kenfinnigan) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:63]: How so?

Ideally we want the methods of LoggingFraction to be chainable on an instance, but we don't need to hold an instance to make that happen.

Maybe I'm not understanding how this is meant to work

**Comment by bobmcwhirter** [comment]
It may also help to fully build the wildfly-swarm-examples repository, to witness what does or doesn't break when you make fundamental changes such as this.

File: logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java (hunk 5 above)
Comment (cmoulliard) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:63]: If you look to the code of the master branch, you will see that an instance of the LogginFraction is created within the static method "createDefaultLoggingFraction" - https://github.com/wildfly-swarm/wildfly-swarm/blob/master/logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java#L121

As now we have a new static method where we will create a defaultFileLoggingFraction, then we should be able to reuse an existing instance of the LoggingFraction instead of creating a new object otherwise when we will chain the 2 static methods within (https://github.com/cmoulliard/wildfly-swarm-core/blob/83ae9cdb1bd3940805e07cf6b705e11a5cfdac8a/testsuite/testsuite-logging/src/test/java/org/wildfly/swarm/logging/FileLoggingInfoLevelArquillianTest.java#L58-L61), then the console will not be created. 

Do you have a better idea to code that ?

Comment (bobmcwhirter) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:63]: I think this is where the hang-up is.

There's two flavors of configuration methods for a given fraction.

The static methods, such as createDebugLoggingFraction() is a helper factory method.  It's a starting point, so you're not starting from scratch with `new LoggingFraction()` and having to do everything by-hand, from nil.

Ultimately, these static methods are doing `new LoggingFraction()`, and then calling a series of _instance_ methods on that object to do work.  Ultimately the fraction is returned to the user (or the container), and no static state is or should be held.  Ever.

When we speak of chaining, that means _instance_ methods, which return _LoggingFraction_ and do work.

Start there.

Once you have friendly instance methods that can be applied to _any_ instance of a `LoggingFraction` object, _then_ you can think about a handy-dandy helper static factory method just to make things easier.

Bottom line:
- Do work in an instance method
- Have that instance method return `this`
- Once you can work with instance methods, _then_ ponder helper factory methods to help people get started for the 80% case.
- Never hold static state. Ever.  Never.  Ever.  It's a smell if you do.

Comment (bobmcwhirter) [reply logging/src/main/java/org/wildfly/swarm/logging/LoggingFraction.java:63]: Basically, if a user starts with

```
LoggingFraction logging = new LoggingFraction()
```

or

```
LoggingFraction logging = LoggingFraction.createDebugLoggingFraction()
```

you should be able work with _either_ of those starting points, allowing for

```
logging.logToFile(...) // or whatever
```

**Comment by heiko-braun** [comment]
If this doesn't make progress I'd suggest we close the PR.

**Comment by kenfinnigan** [comment]
@cmoulliard, are there plans to complete the requested changes?

**Comment by cmoulliard** [comment]
We can delete the PR & close the ticket. I propose that we have a more general discussion later about the conventions to follow to use the FluentApi about the fractions.

**Comment by gastaldi** [comment]
Closing PR for now.
